packages/typespec-ts/src/utils/modelUtils.ts

```diff
@@ -388,37 +388,45 @@ function getSchemaForUnion(
   const variants = Array.from(union.variants.values());
 
   let values = [];
-
+  let namedUnionMember = false;
   for (const variant of variants) {
     // We already know it's not a model type
-    values.push(
-      getSchemaForType(dpgContext, variant.type, { ...options, needRef: false })
-    );
+    const variantType = getSchemaForType(dpgContext, variant.type, {
+      ...options,
+      needRef: false
+    });
+    values.push(variantType);
+    if (variantType.typeName) {
+      namedUnionMember = true;
+    }
   }
-  if (asEnum?.open) {
+  if (asEnum?.open && asEnum.members.size > 0) {
+    namedUnionMember = false;
     values = [];
```

Comment:
just curious if we need to reset the values and namedUnionMember, why we go to foreach in line 392. Should we use open condition to separate them?

Reply:
updated.